simba/optimization.py

```diff
@@ -122,3 +125,269 @@ def service_optimization(schedule, scenario, args):
+                depot_trips[station_name][depot_name] = trip
+        else:
+            # trip from station to depot unknown: save trip as reference
+            depot_trips[station_name][depot_name] = trip
+    else:
+        # no trip to any depot known for this station: save trip as reference
+        depot_trips[station_name] = {depot_name: trip}
+    return depot_trips
+
+
+def generate_depot_trip(station, depot_trips, default_depot_distance, default_mean_speed):
+    """
+    Look up closest depot to a station.
+
+    Assumes default if no trip between this station and a depot exists in the original dataset.
+    :param station: starting location
+    :type station: str
+    :param depot_trips: look-up-table for depot trips
+    :type depot_trips: dict
+    :param default_depot_distance: default assumed average distance from any station to a depot [km]
+    :type default_depot_distance: numeric
+    :param default_mean_speed: default assumed average speed for busses [km/h]
+    :type default_mean_speed: numeric
+    :return: name and distance to closest depot
+    :rtype: dict
+    """
+    try:
+        # get depot with shortest distance
+        depots = sorted(depot_trips[station].items(), key=lambda x: x[1].distance)
+        depot_trip = depots[0]
+        return {
+            "name": depot_trip[0],  # depot station key
+            "distance": depot_trip[1].distance,  # in meters
+            "mean_speed": depot_trip[1].mean_speed,  # in km/h
+            "travel_time": depot_trip[1].arrival_time - depot_trip[1].departure_time,  # timedelta,
+        }
+    except KeyError:
+        # no trip from depot to departure station known: assume defaults
+        assert len(depot_trips) > 0, "No depot trips known"
+        # get any depot key from any station in depot_trips
+        any_station = next(iter(depot_trips.values()))
+        depot_name = next(iter(any_station.keys()))
+        return {
+            "name": depot_name,
+            "distance": default_depot_distance * 1000,
+            "mean_speed": default_mean_speed,
+            "travel_time": datetime.timedelta(hours=default_depot_distance/default_mean_speed),
+        }
+
+
+def recombination(schedule, args, trips, depot_trips):
+    """
+    Recombine individual trips to new rotations.
+
+    Gathers rotations greedy from given trips.
+    Rotations are not mixed, keep trips together if possible.
+    Checks if rotation can be done at all.
+
+    :param schedule: Schedule to be recombined
+    :type schedule: simba.Schedule
+    :param args: Command line arguments
+    :type args: argparse.Namespace
+    :param trips: look-up-table depb rotation ID -> trip list
+    :type trips: dict
+    :param depot_trips: look-up-table for depot trips
+    :type depot_trips: dict
+    :return: recombined schedule
+    :rtype: simba.Schedule
+    """
+    new_rotations = dict()
+    for rot_id, rot in schedule.rotations.items():
+        try:
```

Comment:
iterate over "trips" and get the rotation from schedule rotations instead?

Reply:
Not sure what you mean. The intention here is to leave non-negative and non-depot rotations as they are, but remove negative depot rotations from schedule and replace them with their recombined variants.

Reply:
I mean "trips" from the function arguments. Trips is a dictionary with rotation_ids as values which were prepared for recombination. So they already contain all negative and depot rotations, which need changing. so instead of
new_rotations = dict()
```
    for rot_id, rot in schedule.rotations.items():
        try:
             trip_list = trips[rot_id]
        except KeyError:
            # no info, no change
            new_rotations[rot_id] = rot
            continue
... create new rotation
```

```
new_rotations = deepcopy(schedule.rotations)
    for rot_id, trip_list in trips.items():
       del new_rotations[rot_id)
... create new rotation

```

Reply:
Went even further, don't need new_rotations (work on schedule.rotations instead)